### ISSUE TITLE
Change type definitions / decoder functions to make intellisense more useful

### ISSUE THREAD

**Issue by rommsen** [MangelMaxime/Thot#7]
Example:

The documentation shows the following type:
`list : Decoder<'value> -> Decoder<'value list>`

But if you use the list decoder in your editor, intellisense will show you:
```
val list: 
   decoder: Decoder<'value> ->
   value  : obj             
         -> Result<'value list,DecoderError>
```
This is very confusing, especially if you start with this library or elm style decoders in general.

I think it would be nice, not to use partial application here but to return an actual Decoder explicitely:
```
let list (decoder : Decoder<'value>) : Decoder<'value list> =
    fun (value: obj) ->
        if Helpers.isArray value then
            unbox<obj array> value
            |> Array.map (unwrap decoder)
            |> Array.toList
            |> Ok
        else
            BadPrimitive ("a list", value)
            |> Error
```

What do you think?

**Comment by alfonsogarciacaro** [comment]
Nope, there's no difference. Perhaps surprising, the F# compiler compiles these function exactly in the same way:

```fsharp
let add3 x y z = x + y + z

let add3' x y = fun z -> x + y + z

let add3'' x = fun y z -> x + y + z

let add3''' = fun x y z -> x + y + z
```

[Sample](http://fable.io/repl/#?code=%0Alet%20add3%20x%20y%20z%20%3D%20x%20%2B%20y%20%2B%20z%0A%0Alet%20add3'%20x%20y%20%3D%20fun%20z%20-%3E%20x%20%2B%20y%20%2B%20z%0A%0Alet%20add3''%20x%20%3D%20fun%20y%20z%20-%3E%20x%20%2B%20y%20%2B%20z%0A%0Alet%20add3'''%20%3D%20fun%20x%20y%20z%20-%3E%20x%20%2B%20y%20%2B%20z%0A%0Alet%20add3''''%20%3D%20add3%0A%0Alet%20test()%20%3D%0A%20%20let%20x%20%3D%20add3%201%202%203%0A%20%20let%20y%20%3D%20add3'%201%202%203%0A%20%20let%20f1%20%3D%20add3%201%202%0A%20%20let%20f2%20%3D%20add3'%201%202%0A%20%20f1%203%20%2B%20f2%203%0A%20%20%0A&astInput=false&babili=false&evaluate=false&lineWrap=false&presets=es2015%2Creact%2Cstage-2)

It's only when there are no arguments that it's compile differently `let add3'''' = add3`

**Comment by MangelMaxime** [comment]
I will fix this issue after merging #4 as it will lead to a lot of diff :)

**Comment by MangelMaxime** [comment]
@rommsen I just released version 0.1.0-beta-003 with better type definitions. I wasn't able to change all of them because I didn't know how to code it :) But the primary are using `Decoder` for the signature. 

If you know how to fix the missing one I am open for a PR :) thanks for the suggestion
